lib/Parse/Lexer.cpp

```diff
@@ -1951,36 +1958,31 @@ const char *Lexer::findEndOfCurlyQuoteStringLiteral(const char *Body,
   }
 }
 
-void Lexer::lexRegexLiteral(const char *TokStart) {
+bool Lexer::tryLexRegexLiteral(const char *TokStart) {
   assert(*TokStart == '\'');
 
-  bool HadError = false;
-  while (true) {
-    // Check if we reached the end of the literal without terminating.
-    if (CurPtr >= BufferEnd || *CurPtr == '\n' || *CurPtr == '\r') {
-      diagnose(TokStart, diag::lex_unterminated_regex);
-      return formToken(tok::unknown, TokStart);
-    }
+  // We need to have experimental string processing enabled, and have the
+  // parsing logic for regex literals available.
+  if (!LangOpts.EnableExperimentalStringProcessing || !regexLiteralLexingFn)
+    return false;
 
-    const auto *CharStart = CurPtr;
-    uint32_t CharValue = validateUTF8CharacterAndAdvance(CurPtr, BufferEnd);
-    if (CharValue == ~0U) {
-      diagnose(CharStart, diag::lex_invalid_utf8);
-      HadError = true;
-      continue;
-    }
-    if (CharValue == '\\' && (*CurPtr == '\'' || *CurPtr == '\\')) {
-      // Skip escaped delimiter or \.
-      CurPtr++;
-    } else if (CharValue == '\'') {
-      // End of literal, stop.
-      break;
-    }
-  }
-  if (HadError)
-    return formToken(tok::unknown, TokStart);
+  // Ask libswift to try and lex a regex literal, and bail if it can't.
+  const char *ErrStr = nullptr;
+  auto *EndPtr = regexLiteralLexingFn(TokStart, BufferEnd, &ErrStr);
+  if (!EndPtr)
+    return false;
 
+  // Update to point to where we ended regex lexing.
+  assert(EndPtr > TokStart && EndPtr <= BufferEnd);
+  CurPtr = EndPtr;
+
+  if (ErrStr) {
+    diagnose(TokStart, diag::regex_literal_parsing_error, ErrStr);
+    formToken(tok::unknown, TokStart);
+    return true;
+  }
```

Comment:
Could you add an assertion checking `CurPtr[-1]` is a valid delimiter character?